clients/pkg/promtail/targets/docker/target.go

```diff
@@ -0,0 +1,209 @@
+		metrics:       metrics,
+
+		cancel:  cancel,
+		client:  client,
+		running: atomic.NewBool(false),
+	}
+	go t.processLoop(ctx)
+	return t, nil
+}
+
+func (t *Target) processLoop(ctx context.Context) {
+	t.running.Store(true)
+
+	opts := docker_types.ContainerLogsOptions{
+		ShowStdout: true,
+		ShowStderr: true,
+		Follow:     true,
+		Timestamps: true,
+		Since:      strconv.FormatInt(t.since, 10),
+	}
+
+	logs, err := t.client.ContainerLogs(ctx, t.containerName, opts)
+	if err != nil {
+		level.Error(t.logger).Log("msg", "could not fetch logs for container", "container", t.containerName, "err", err)
+		t.err = err
+		return
+	}
+
+	// Start transferring
+	rstdout, wstdout := io.Pipe()
+	rstderr, wstderr := io.Pipe()
+	t.wg.Add(1)
+	go func() {
+		defer func() {
+			t.wg.Done()
+			wstdout.Close()
+			wstderr.Close()
+			t.Stop()
+		}()
+
+		written, err := stdcopy.StdCopy(wstdout, wstderr, logs)
+		if err != nil {
+			level.Warn(t.logger).Log("msg", "could not transfer logs", "written", written, "container", t.containerName, "err", err)
+		} else {
+			level.Info(t.logger).Log("msg", "finished transferring logs", "written", written, "container", t.containerName)
+		}
+	}()
+
+	// Start processing
+	t.wg.Add(2)
+	go t.process(rstdout, "stdout")
+	go t.process(rstderr, "stderr")
+
+	// Wait until done
+	<-ctx.Done()
+	t.running.Store(false)
+	logs.Close()
+	level.Debug(t.logger).Log("msg", "done processing Docker logs", "container", t.containerName)
+}
+
+// extractTs tries for read the timestamp from the beginning of the log line.
+// It's expected to follow the format 2006-01-02T15:04:05.999999999Z07:00.
+func extractTs(line string) (time.Time, string, error) {
+	pair := strings.SplitN(line, " ", 2)
+	if len(pair) != 2 {
+		return time.Now(), line, fmt.Errorf("Could not find timestamp in '%s'", line)
+	}
+	ts, err := time.Parse("2006-01-02T15:04:05.999999999Z07:00", pair[0])
+	if err != nil {
+		return time.Now(), line, fmt.Errorf("Could not parse timestamp from '%s': %w", pair[0], err)
+	}
+	return ts, pair[1], nil
+}
+
+func (t *Target) process(r io.Reader, logStream string) {
+	defer func() {
+		t.wg.Done()
+	}()
+
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		line := scanner.Text()
+		ts, line, err := extractTs(line)
+		if err != nil {
+			level.Error(t.logger).Log("msg", "could not extract timestamp, skipping line", "err", err)
+			t.metrics.dockerErrors.Inc()
+			continue
+		}
+
+		t.handler.Chan() <- api.Entry{
+			Labels: t.labels.Merge(model.LabelSet{dockerLabelLogStream: model.LabelValue(logStream)}),
+			Entry: logproto.Entry{
+				Timestamp: ts,
+				Line:      line,
+			},
+		}
+		t.metrics.dockerEntries.Inc()
+		t.positions.Put(positions.CursorKey(t.containerName), ts.Unix())
+	}
+
+	err := scanner.Err()
+	if err != nil {
+		level.Warn(t.logger).Log("msg", "finished scanning logs lines with an error", "err", err)
+	}
+
+}
+
+func (t *Target) Stop() {
+	t.cancel()
+	t.wg.Wait()
+	t.handler.Stop()
```

Comment:
I think you have a bug here. It seems to me that you stop the handler but it's a shared one. You either need to create one pipeline stage per target or close it globally just once.